### PR TITLE
Migrate datetime chooser init script to Stimulus `w-date`

### LINKED ISSUE #9406: Dateinput widget does not work in Custom Document modal chooser
I create a custom document model with a custom datefield named "record_date". Its default value is the current date.
HomePage has a field named "doc" which is referenced by the custom document.

The problem is that when I click the "record_date" field in the document modal chooser, the datepicker widget does not show.
image

However, when I edit the document in document edition page, the datepicker shows properly.

Wagtail version: 4.0.2
django version: 4.1.2
python version: 3.10

### LINKED ISSUE #10260: 🎛️ Migrate datetime picker widgets to a Stimulus Controller `w-date`
> ℹ️ **Part of the [Stimulus 🎛️ RFC 78](https://github.com/wagtail/rfcs/pull/78)**

### Is your proposal related to a problem?

After reviewing #10255 & #10254 - it seems to make sense to try to reduce the amount of extra code needed to support date widgets in any part of the Wagtail admin.

It would be good to refresh the current Date Picker widget code and provide a future compatible way to support either no widget OR non-jQuery widget approaches.

#### Current implementation

- [`client/src/entrypoints/admin/date-time-chooser.js`](https://github.com/wagtail/wagtail/blob/main/client/src/entrypoints/admin/date-time-chooser.js)
- [`wagtail/admin/widgets/datetime.py`](https://github.com/wagtail/wagtail/blob/main/wagtail/admin/widgets/datetime.py)
- [`client/src/entrypoints/admin/telepath/widgets.js`](https://github.com/wagtail/wagtail/blob/main/client/src/entrypoints/admin/telepath/widgets.js#L400-L453)

#### Proposed implementation

A new, single, controller `w-date`/`DateController` that accepts three `mode`s (`date`,`datetime` and `time`).

Replace the existing implementation almost verbatim with Stimulus doing the initialisation instead of a global function call.

With the following refinements;

- Avoid the `focus({ soft: true })` approach in StreamField and instead support a nicer work around that honours keyboard accessibility.
- Pull out the `dateEqual` function to its own util and add tests (plus avoid using the deprecated `getYear` function).
- Must have unit tests that cover basic scenarios.
- Remove the need for checking the global `dateTimePickerTranslations` for custom widget options, this is approach has not been used for nearly 9 years. See 5936c020d8099e8d3f8a54df1f2ef6520296d938 & 9f33ef313781eeaa5857e767092c9950224a583c
- Ensure that when the widget's DOM gets removed, the jQuery widget gets removed (destroyed) to avoid risk of memory issues when StreamField or InlinePanel is used a lot.

#### Validation check-list

- [ ] InlinePanel usage with one of the date picker variants (bakerydemo bakery page model)
- [ ] StreamField block usage with variants of the date widgets (add some to `bakerydemo/base/blocks.py` in bakerydemo)
- [ ] Date picker widgets used in reporting and ModelAdmin field listings

### Describe alternatives you've considered

- This code is not compliant for CSP, it should be reviewed at some point.

### Additional context

- This is part of a larger goal to migrate to Stimulus, see RFC 78 linked above
- Potentially will fix #9406
- Will make it easier to support #8056 while providing a clean way for others to include their own date picker widgets if they choose
- Will make it easier to support customisations to work around #4768 (with a custom date picker implementation)
- Could improve some of the keyboard control issues when using date pickers in StreamField #5325

### PR DESCRIPTION
**Pull request by lb-**
- Avoid all need for Telepath widgets for datetimepicker, remove all inline script usage, replace with a single Stimulus controller 'w-date'
- Refines 'soft' focus approach (which just did not focus on datetimepicker at all) with a DOM event / preventDefault approach so that we can use similar functionality without needing the overhead of a Telepath instance for just this behaviour. Add unit tests for this new Telepath widget behaviour
- Ensure that Block variants use the same HTML (which will now include autocomplete off)
- Ensure that options cannot override core functionality (hideCurrent, change event dispatch)
- Split out date compare into own util (with tests)
- Add unit test functionality to cover all datetimepicker usage scenarios
- Remove the need for a dedicated datetimepicker JS import
- Support existing wagtailConfig global usage
- Fix console error when running widget unit tests (global scrollTo not provided by JSDom)
- Find another test case to check the media inheritance approach as no longer needed for Admin date widgets
- Fixes #10260

## Implementation & testing notes

- Tested on Safari 16 (a very small flash of the widget shows in Safari), Firefox 101, Chrome 111. Tested with macOS voiceover on all browsers.
- The soft focus approach works better than before with making the actual date picker the 'next' keyboard focusable target and correctly reads the context of the container to screen readers.

### CODE REVIEW

File: wagtail/admin/tests/test_widgets.py [hunk 1]
```diff
@@ -165,213 +166,151 @@ def test_render_js_init_with_can_choose_root(self):
...
+class TestAdminTaskChooser(TestCase):
+    def test_media_inheritance(self):
+        """
+        Widgets inheriting from AdminTaskChooser should have their media definitions merged
+        with AdminTaskChooser's
+        """
+        widget = AdminStarTaskChooser()
+        media_html = str(widget.media)
+        self.assertIn("wagtailadmin/js/task-chooser.js", media_html)
+        self.assertIn("vendor/star_date.js", media_html)
```
Comment (lb-) [inline wagtail/admin/tests/test_widgets.py:178]:
```suggestion
        self.assertIn("vendor/star_tasks.js", media_html)
```

File: client/src/controllers/DateController.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  get options() {
+    return {
+      // merging default options based on mode and global
+      ...(this.constructor.defaultOptions[this.modeValue] || {}),
+      // allow provided options to override any config
+      ...this.optionsValue,
+      // ensure callbacks cannot be overridden
+      onGenerate: this.hideCurrent,
+      onChangeDateTime: () => {
+        this.dispatch('change', { prefix: '' });
```
Comment (lb-) [inline client/src/controllers/DateController.ts:94]:
```suggestion
        this.dispatch('change', { prefix: '', cancelable: false });
```

File: client/src/controllers/DateController.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  get options() {
+    return {
+      // merging default options based on mode and global
+      ...(this.constructor.defaultOptions[this.modeValue] || {}),
+      // allow provided options to override any config
+      ...this.optionsValue,
+      // ensure callbacks cannot be overridden
+      onGenerate: this.hideCurrent,
+      onChangeDateTime: () => {
+        this.dispatch('change', { prefix: '', cancelable: false });
```
Comment (lb-) [inline client/src/controllers/DateController.ts:94]: Should add a unit test for this

**Comment by lb-** [comment]
This is now ready for review.

**Comment by lb-** [comment]
Rebased with upgrade considerations moved to 5.1 release notes.

**Comment by lb-** [comment]
Fixed up the additional failing tests - should be good now.

**Comment by lb-** [comment]
Rebased

**Comment by lb-** [comment]
This has not had a review yet but I have been thinking about the previous approach and found a much simpler solution for the 'soft' focus for use in the date picker.

This now avoids any <100ms flash of the datepicker being open in Safari and overall is a cleaner solution in the code.

I also reverted the removal of the soft focus call as this is a documented approach. I have opted to NOT document the new Custom Event dispatching as it will be for default widgets only and is not something that can be easily implemented at the top level usage of calling `focus`.

Would love to see this get a review when someone has a chance, I feel this is a more accessible/secure, simpler and hopefully more maintainable solution. Of course happy to take feedback and push back on the implementation.

**Review by laymonage — COMMENTED** [review]
Hey @lb-, thank you for putting this together! I had a chat with Thibaud (and if I remember correctly the accessibility subteam had discussed this as well), we're keen on migrating the date/time pickers to the browser's native inputs. We're not exactly sure if everything can be handled gracefully with the browser's built-in inputs (especially around timezones), so we might still need some JavaScript via Stimulus in the end. In any case, this means we won't be needing the xdsoft library, which unfortunately means we won't go ahead with this PR as-is. I hope that's OK!

**Comment by lb-** [comment]
no problems, I knew that was coming at some but but still assumed we would want the basics in place to hook in the JS changes/reduction when it came.

I do think there would be a desire from those working with Wagtail to opt for custom date widgets and assumed we would likely want to support that in some basic way.

If we are close to implementing the new approach, sounds good. If we feel it's still a release or two away, I would like to suggest we do review this PR for merging as it is one of the largest sources of inline script tags.

Happy with whatever we do here. Feel free to close this PR if it's not something we want to merge at this time though.

**Comment by lb-** [comment]
@laymonage & @thibaudcolas - a proposal for a different tact for date pickers, would love your thoughts.

1.  move ahead with this PR, I can review and revise based on the cancelled feedback. 
2. I add a new setting, something like `wagtail_legacy_date_fields`, this will be set to True for 5.1 with release notes about this config, ready to change the default to False in a future release. 
3. I adjust the JS to load Async, dynamic require, only if the controller loads on a page and the legacy approach is true.

This gives us the benefits of being able to pull out critical path inline scripts, gives us time to confirm the native field approach and also paves the way for being closer to removing insert_editor_js. 

We can set this config to False in bakerydemo, Guide site and maybe the Wagtail website. Plus gives some in the community to trial the approach.

If we realise we do need some JS, we have the Controller scaffold to build on. If we simply want to make it easier to BYO date widgets, we can change the controller to be a basic skeleton.

I'm probably a bit biased, having worked on this for a while, but do want to push a bit on giving us all time to fully move away from the xdsoft code but also give us a nice win for CSP, JS loading and accessibility options.

**Comment by lb-** [comment]
Rebased/resolved conflicts.

**Comment by laymonage** [comment]
@lb- Thibaud and I have gone back and forth on this a few times, and I think we were still leaning towards completely replacing the inputs to the native browser inputs and making sure that we iron out any issues before merging, as opposed to having the old one as fallback in case there are issues with the new implementation. Even without a fallback, reviewing this PR is quite non-trivial and since we're going to replace it soon anyway, we feel it'd be better to focus our time on the new implementation instead.

It does sound a bit too optimistic to think that we'll be able to cover any issues before the change is released, though. That said, we still have the option to release bug fixes, so I think some friction in there is worth compared to maintaining two different implementations plus the code to handle the switch between the two.

I'm planning to pick the new implementation up in 5.2, and in the mean time I think you could hold off from continuing any work on this PR. Sorry we didn't get to review this, but hopefully the change is for the better.

**Comment by lb-** [comment]
@laymonage thank you, I'm more than happy to align with this direction. Thanks for getting back to me on this.

I'll close this PR and the related isssue and add some notes where needed.

**Comment by lb-** [comment]
Closing as per comment above, we will not be taking this approach for the date pickers.

**Pull request closed without merging**